core/src/main/java/io/grpc/internal/TransportSet2.java

```diff
@@ -0,0 +1,495 @@
+    this.transportFactory = transportFactory;
+    this.scheduledExecutor = scheduledExecutor;
+    this.connectingTimer = stopwatchSupplier.get();
+    this.channelExecutor = channelExecutor;
+    this.callback = callback;
+  }
+
+  /**
+   * Returns a READY transport that will be used to create new streams.
+   *
+   * <p>Returns {@code null} if the state is not READY.
+   */
+  @Nullable
+  final ClientTransport obtainActiveTransport() {
+    ClientTransport savedTransport = activeTransport;
+    if (savedTransport != null) {
+      return savedTransport;
+    }
+    Runnable runnable = null;
+    synchronized (lock) {
+      // Check again, since it could have changed before acquiring the lock
+      if (activeTransport != null) {
+        return activeTransport;
+      }
+      if (state.getState() == IDLE) {
+        gotoNonErrorState(CONNECTING);
+        runnable = startNewTransport();
+      }
+    }
+    if (runnable != null) {
+      runnable.run();
+    }
+    return null;
+  }
+
+  @CheckReturnValue
+  @GuardedBy("lock")
+  private Runnable startNewTransport() {
+    Preconditions.checkState(reconnectTask == null, "Should have no reconnectTask scheduled");
+
+    if (nextAddressIndex == 0) {
+      connectingTimer.reset().start();
+    }
+    List<SocketAddress> addrs = addressGroup.getAddresses();
+    final SocketAddress address = addrs.get(nextAddressIndex++);
+    if (nextAddressIndex >= addrs.size()) {
+      nextAddressIndex = 0;
+    }
+
+    ConnectionClientTransport transport =
+        transportFactory.newClientTransport(address, authority, userAgent);
+    if (log.isLoggable(Level.FINE)) {
+      log.log(Level.FINE, "[{0}] Created {1} for {2}",
+          new Object[] {getLogId(), transport.getLogId(), address});
+    }
+    pendingTransport = transport;
+    transports.add(transport);
+    return transport.start(new TransportListener(transport, address));
+  }
+
+  /**
+   * Only called after all addresses attempted and failed (TRANSIENT_FAILURE).
+   * @param status the causal status when the channel begins transition to
+   *     TRANSIENT_FAILURE.
+   */
+  @GuardedBy("lock")
+  private void scheduleBackoff(final Status status) {
+    class EndOfCurrentBackoff implements Runnable {
+      @Override
+      public void run() {
+        try {
+          Runnable runnable = null;
+          synchronized (lock) {
+            reconnectTask = null;
+            if (state.getState() == SHUTDOWN) {
+              // Even though shutdown() will cancel this task, the task may have already started
+              // when it's being cancelled.
+              return;
+            }
+            gotoNonErrorState(CONNECTING);
+            runnable = startNewTransport();
+          }
+          if (runnable != null) {
+            runnable.run();
+          }
+        } catch (Throwable t) {
+          log.log(Level.WARNING, "Exception handling end of backoff", t);
+        }
+      }
+    }
+
+    if (state.getState() == SHUTDOWN) {
```

Comment:
Reorganized the caller so that checkArgument(CONNECTED) is called prior to this method.